pyhms/utils/covariance_estimate.py

```diff
@@ -0,0 +1,71 @@
+import numpy as np
+from leap_ec import Individual
+
+from ..demes.abstract_deme import AbstractDeme
+
+N_INDIVIDUALS_PER_DIMENSION = 25
+_EPS = 1e-10
+
+
+def find_closest_rows(X: np.ndarray, y: np.ndarray, top_n: int) -> np.ndarray:
+    distances = np.sqrt(((X - y) ** 2).sum(axis=1))
+    closest_indices = np.argsort(distances)
+    top_indices = closest_indices[:top_n]
+    return X[top_indices]
+
+
+def estimate_covariance(X: np.ndarray) -> np.ndarray:
+    return np.cov(X.T, bias=1)  # type: ignore[call-overload]
+
+
+def estimate_sigma0(X: np.ndarray) -> float:
+    cov_estimate = estimate_covariance(X)
+    sigma0 = np.sqrt(np.trace(cov_estimate) / len(cov_estimate))
+    # Avoid ZeroDivisionError in cma.CMAEvolutionStrategy.
+    return max(sigma0, _EPS)
+
+
+def estimate_stds(X: np.ndarray) -> np.ndarray:
+    return np.sqrt(np.diag(estimate_covariance(X)))
+
+
+def get_population(
+    parent_deme: AbstractDeme,
+    x0: Individual,
+    use_closest_rows: bool | None = True,
+) -> np.ndarray:
+    n_individuals = N_INDIVIDUALS_PER_DIMENSION * len(x0.genome)
+    parent_population = np.array([ind.genome for pop in parent_deme.history for ind in pop])
+    if use_closest_rows:
+        population = find_closest_rows(parent_population, x0.genome, n_individuals)
+    else:
+        population = parent_population[-n_individuals:]
+    return population
+
+
+def get_initial_sigma0(
+    parent_deme: AbstractDeme,
+    x0: Individual,
+    use_closest_rows: bool | None = True,
+) -> float:
+    population = get_population(parent_deme, x0, use_closest_rows)
+    return estimate_sigma0(population)
+
+
+def get_initial_stds(
+    parent_deme: AbstractDeme,
+    x0: Individual,
+    use_closest_rows: bool | None = True,
+) -> np.ndarray:
+    population = get_population(parent_deme, x0, use_closest_rows)
+    return estimate_stds(population)
+
+
+def get_initial_sigma0_from_bounds(bounds: np.ndarray) -> float:
```

Comment:
Regarding this type of sigma0 setting, it is a basic approach to determine a good guess for sigma0 and can be used as a default, although I'm not sure, if it is viable in our application. At this point I find the only usage in tests, so I'm not sure about the context - would the bounds correspond to the original problem? According to pycma:
"""
The problem variables should have been scaled, such that a single standard deviation on all variables is useful and the optimum is expected to lie within about `x0` +- ``3*sigma0``. Often one wants to check for solutions close to the initial point. This allows, for example, for an easier check of consistency of the objective function and its interfacing with the optimizer. In this case, a much smaller `sigma0` is advisable.
"""
Optunas approach uses a minimal diameter instead of scaling and does not make assumptions about the location of optimum - thus if you start at the center, then you take diameter/6 according to the quote above (although without scaling it may be useless for some of the variables). In our case CMA-ES is used on lower levels so we do not start at the center of domain and we have some knowledge about the possible location of optima. Even if the user lacks knowledge and sets the sigma0 according to our bound-based method, for the lower levels a smaller sigma would be more appropriate. The existence of this method is not a problem, but I'm not sure how much it adds. I leave a decision to you.

Reply:
Good catch. I decided to delete this method right now. The idea was to have a function to estimate sigma0 from the problem definition (without any additional computational cost).